homeassistant/components/sensor/history_average.py

```diff
@@ -0,0 +1,342 @@
+                break
+            else:
+                # remove if the next item is also before start
+                next_index = index + 1
+                if (next_index < len(self._history) and
+                        (not after_start(self._history[next_index],
+                                         start_timestamp))):
+                    del self._history[index]
+
+    def get_period(self):
+        """Return current period, used for testing."""
+        return self._period
+
+    def __init__(
+            self, hass, entity_id, start, end, duration,
+            name, unit):
+        """Initialize the HistoryAverage sensor."""
+        self._hass = hass
+        self._entity_id = entity_id
+        self._duration = duration
+        self._start = start
+        self._end = end
+        self._name = name
+        self._unit_of_measurement = unit
+
+        # set defaults at init
+        now = dt_util.utcnow()
+        self._period = (now, now)
+        self._history = []
+        self._state = 0
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Initialize data and register callbacks."""
+        # load data from recorder and do initial calculation
+        yield from self.async_get_history()
+        yield from self.async_update()
+
+        @callback
+        def state_listener(entity, old_state, new_state):
+            """Handle the sensor state changes."""
+            self._history.append(new_state)
+            sorted(self._history, key=attrgetter('last_changed'))
+            self._hass.async_add_job(self.async_update_ha_state(True))
+
+        @callback
+        def sensor_startup(event):
+            """Update on startup."""
+            async_track_state_change(
+                self._hass, self._entity_id, state_listener)
+
+            self._hass.async_add_job(self.async_update_ha_state(True))
+
+        self._hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, sensor_startup)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the state is expressed in."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        # Since the values can/do change with time, not just with
+        # state changes, the sensor should poll
+        return True
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        def period_in_seconds(period):
+            """Get the period period duration in seconds."""
+            if len(period) != 2 or period[0] == period[1]:
+                return 0.0
+            return (period[1] - period[0]).total_seconds()
+
+        def pretty_duration(period):

```

Comment:
We should not have a pretty duration. Attributes should be machine readable. If pretty printend is necessary, the frontend should add this.

Reply:
I agree. However, I'm not sure what data type to return here to represent a length in time. I tried `timedelta`, but it's not JSON serializable. Do you have a suggestion?